lib/opentok/archives.rb

```diff
@@ -226,5 +226,102 @@ def layout(archive_id, options = {})
       response = @client.layout_archive(archive_id, options)
       (200..300).include? response.code
     end
+
+    # Adds a stream to currently running composed archive that was started with the
+    # streamMode set to "manual". For a description of the feature, see
+    # {https://tokbox.com/developer/rest/#selecting-archive-streams}.
+    #
+    # @param [String] archive_id
+    #   The archive ID.
+    #
+    # @param [String] stream_mode
+    #   The stream mode of the archive. Must be "manual" in order to add a stream.
+    #
+    # @option opts [String] :add_stream
+    #   The ID for the stream to be added to the archive
+    #
+    # @option opts [true, false] :has_audio
+    #   (Boolean, optional) — Whether the composed archive should include the stream's
+    #   audio (true, the default) or not (false).
+    #
+    # @option opts [true, false] :has_video
+    #   (Boolean, optional) — Whether the composed archive should include the stream's
+    #   video (true, the default) or not (false).
+    #
+    # You can call the method repeatedly with add_stream set to the same stream ID, to
+    # toggle the stream's audio or video in the archive. If you set both has_audio and
+    # has_video to false, you will get error response.
+    #
+    # @raise [ArgumentError]
+    #   The archive_id parameter is empty.
+    #
+    # @raise [ArgumentError]
+    #   The streamMode for the archive is not set to "manual".
+    #
+    # @raise [ArgumentError]
+    #   The options parameter is empty.
+    #
+    # @raise [ArgumentError]
+    #   The add_stream property of the options parameter is empty.
+    #
+    # @raise [ArgumentError]
+    #   The has_audio and has_video properties of the options parameter are both set to "false"
+    #
+    def add_stream(archive_id, stream_mode, options)
```

Comment:
Please add `stream_mode` to the `options` of the `opentok.archives.create` method and as an attribute of the Archive class. (The same comments apply to broadcast.)

Reply:
> Please remove the stream_mode parameter. You do not set the stream mode when calling this method (you set it when you start the archive).

`Archives#add_stream` isn't intended to be called directly, only via the `Archive#add_stream` [wrapper method](https://github.com/opentok/OpenTok-Ruby-SDK/blob/c4a711ea9fe58725b40622e99efb92b3f1f80f71/lib/opentok/archive.rb#L157); this method passes in whatever the stream mode is for the currently running archive. The only reason I was passing this into `Archives#add_stream` was to perform a check on it there, but I can move the check to the `Archive#add_stream` wrapper method instead and not pass the argument (I only did it this way to try and keep the wrapper methods as simple as possible in terms of logic).

> The Archive class should include a stream_mode attribute and a streams attribute. (The same comments apply to broadcast.)

I'm not entirely sure what you mean here. If you mean store the `stream_mode` data somewhere on the `Archive` object itself, it's already stored in the Hash assigned to the `@json` [instance variable](https://github.com/opentok/OpenTok-Ruby-SDK/blob/c4a711ea9fe58725b40622e99efb92b3f1f80f71/lib/opentok/archive.rb#L71). TBH, I don't think this is an optimal approach, but I also don't particularly want to go refactoring parts of the current implementation that aren't directly related to adding this new feature.

Reply:
Re adding stream_mode attribute and a streams attributes to the Archive and Broadcast classes, I suggested it because the Archive and Broadcast objects returned by the REST methods now include `streamMode` and `streams` propperties. (See https://tokbox.com/developer/rest/#retrieve_archive_info and https://tokbox.com/developer/rest/#get_info_broadcast). Even if these properties are exposed via the @json instance variable, shouldn't we add `@attr` references to these in the docs comments (`@attr [Array] streams` and `@attr [string] streamMode`)?

Reply:
Hi @jeffswartz. I'm not opposed to adding properties on the objects (FWIW, I don't particularly like the current implementation of using a Hash assigned to the `@json` ivar). However, I'd suggest making such changes as part of a separate PR (or even waiting until we roll the video functionality into the core SDK next year -- there's also a bunch of other stuff I'd like to change about the way these classes are implemented), since a number of other methods not related to this particular feature would also need to be refactored.

Regarding the docs comments, the data in the hash assigned to the `@json` instance variable in the `initialize` method represents the response from starting an Archive/ Broadcast. 
- For `broadcast.rb` the comments seem to be in line with the [documentation](https://tokbox.com/developer/rest/#start_broadcast), including the [`streamMode` comment](https://github.com/opentok/OpenTok-Ruby-SDK/blob/9eebe5f83db4e81669d6b067c3ea46c3e45597cd/lib/opentok/broadcast.rb#L25); `streams` isn't listed in the comments, but this isn't in the docs either.
- For `archive.rb`, again the [`streamMode` comment](https://github.com/opentok/OpenTok-Ruby-SDK/blob/9eebe5f83db4e81669d6b067c3ea46c3e45597cd/lib/opentok/archive.rb#L42) is there, though not `streams` (though again this isn't listed in the [documentation](https://tokbox.com/developer/rest/#start_archive) for the response data for starting an archive). There are a couple of discrepancies with the documentation (which I will fix): `partner_id` should be `projectId`, and `resolution` is missing.

Reply:
Updated the `Archive` comments in [this commit](https://github.com/opentok/OpenTok-Ruby-SDK/pull/235/commits/e14c4e4db86d9743ba735994e86a6807d1cff826)